### PR TITLE
Fix telemetry shutdown in the case of a failed config update.

### PR DESCRIPTION
**Pull request by BrynCooke**
The root cause is that tracing providers MUST be flushed before drop. otherwise they hang. This is really nasty, but the fix needs to happen in Otel.

Flushing will happen automagically if a provider is set as the global for otel.

In the case of a plugin after the otel plugin failing to start up correctly, or even one of the subsequent methods inside the otel plugin fails then the provider is not set as the global and we must flush it.

Flushing must occur in a separate thread.

Flush will now happen when a telemetry plugin is dropped.

### CODE REVIEW

**Review by bnjjj — APPROVED** [review]

**Comment by o0Ignition0o** [comment]
Does this PR supersede #848 ? I'm about to add a test to it, shall i add it here instead?

**Comment by BrynCooke** [comment]
> 

Yes

**Comment by BrynCooke** [comment]
One of the state machine tests needs updating

**Comment by BrynCooke** [comment]
As a follow up it'd be good to get rid of startup and shutdown

**Comment by o0Ignition0o** [comment]
I have added a regression test for the hang bug, which can be found [in this commit](https://github.com/apollographql/router/pull/848/commits/eabc41dea50a8898e3e911ff586ecfcfb46e23b9)

running this command:

```bash
cargo test --package apollo-router --lib -- router_factory::test::test_telemetry_doesnt_hang_with_invalid_schema --exact --nocapture
```

shall pass the test, while it hangs here (and on main)

**Comment by BrynCooke** [comment]
@o0Ignition0o Your test showed that my approach was flawed. Pushed a solution that relies only on Drop.

File: apollo-router/src/testdata/invalid_supergraph.graphql [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# This supergraph is almost valid. the only difference is that the reviews field in the Product type joins on REVIEWS which is incorrect
```
Comment (o0Ignition0o) [inline apollo-router/src/testdata/invalid_supergraph.graphql:1]:
```suggestion
# This supergraph is almost valid. the only difference is that the reviews field in the Product type joins on PRODUCTS which is incorrect
```

My bad 😅

**Pull request merged**